nx/lib/nx/binary_backend.ex

```diff
@@ -1082,6 +1083,186 @@ defmodule Nx.BinaryBackend do
+        # Execute a round of jacobi rotations on u, d and v
+        {u_bin, d_bin, v_bin} =
+          for p <- 0..(m - 1), q <- (p + 1)..(n - 1), reduce: {u_bin, d_bin, v_bin} do
+            {u_bin, d_bin, v_bin} ->
+              # We need the values for d_bin at indices [p,p], [p,q], [q,p], [q,q], [[p,q], 0..n-1] for this first iteration
+              # However, we'll also need to modify d_bin afterwards, so we need to both access these values and accumulate
+              # the bitstrings that come before, between and after, [p, 0..n-1] and [q, 0..n-1] for updating afterwards
+              d_values =
+                match_types [output_type] do
+                  acc = %{
+                    d_pp: "",
+                    d_pq: "",
+                    d_qp: "",
+                    d_qq: "",
+                    row_p: "",
+                    row_q: "",
+                    before_row_p: "",
+                    between_rows_p_and_q: "",
+                    after_row_q: "",
+                    row: 0,
+                    col: 0
+                  }
+
+                  for <<match!(x, 0) <- d_bin>>, reduce: acc do
+                    acc ->
+                      acc =
+                        case {row, col} do
+                          {^p, ^p} ->
+                            acc
+                            |> Map.put(:d_pp, x)
+                            |> Map.put(:row_p, acc.row_p <> x)
+
+                          {^p, ^q} ->
+                            acc
+                            |> Map.put(:d_pq, x)
+                            |> Map.put(:row_p, acc.row_p <> x)
+
+                          {^p, _} ->
+                            Map.put(acc, :row_q, acc.row_p <> x)
+
+                          {^q, ^q} ->
+                            acc
+                            |> Map.put(:d_qq, x)
+                            |> Map.put(:row_q, acc.row_q <> x)
+
+                          {^q, ^p} ->
+                            acc
+                            |> Map.put(:d_qp, x)
+                            |> Map.put(:row_q, acc.row_q <> x)
+
+                          {^q, _} ->
+                            Map.put(acc, :row_q, acc.row_q <> x)
+
+                          {row, _} when row < p ->
+                            Map.put(acc, :before_row_p, acc.before_row_p <> x)
+
+                          {row, _} when row > p and row < q ->
+                            Map.put(acc, :before_row_p, acc.between_rows_p_and_q <> x)
+
+                          _ ->
+                            Map.put(acc, :after_row_q, acc.after_row_q <> x)
+                        end
+
+                      {row, col} =
+                        if col + 1 == n do
+                          {row + 1, 0}
+                        else
+                          {row, col + 1}
+                        end
+
+                      acc |> Map.put(:row, row) |> Map.put(:col, col)
+                  end
+                end
```

Comment:
The copying of binary here in a loop is going to be expensive (i.e. `acc.before_row_p <> x`). My suggestion is to move this to a list (or a list of lists) and provide a version of `dot_binary` that works on lists. Since it is always MxN, perhaps a hardcoded version will suffice. :) What do you think?

Reply:
Right, I'll think about this. One of the advantages is being able to keep everything in the numerical-type domain, serializing to binary only at the end. This would remove the need for repeated `match_types` + `for` constructs.

One concern I have is about needing to access columns and being able to replace them as well. One idea I was thinking about for random access but probably performs really poorly is having a `%{linear_index => element}` map, which could be indexed through `linear_index = row * num_cols + col`. Another problem with this approach is that I don't think that `Map.values` would be realiable for retrieving the matrix as a whole due to the fact that ordering is not guaranteed.

A list of lists works fine for reading and replacing rows at O(num_rows), assuming a list of rows, but ends up being O(num_rows * num_cols) for accessing columns.

If we go with the "list of lists" approach, I'm thinking of accessing elements using constructs like:

```elixir
#matrix[[row, 0..num_cols-1]]
row = Enum.at(matrix, row)

#matrix[[row, col]]
element = matrix |> Enum.at(row) |> Enum.at(col)

# access matrix[[0..num_rows-1, col]]
column = matrix |> List.flatten() |> Enum.drop(col) |> Enum.take_every(num_cols)

# For lists of columns maybe a `reduce` over `List.flatten(matrix)` is probably better than the following:
# matrix[0..num_rows-1, cols]

flat_matrix = List.flatten(matrix)
columns = Enum.map(cols, & &1 |> Enum.drop() |> Enum.take_every(num_cols))
```

For replacing rows, it's pretty straight forward, but for replacing columns and elements, it does sound like a nested loop of some sort will be needed.

Any thoughts on this?

Reply:
For lists of lists, could you transpose the dimensions whenever you need to access it by columns? then the dot should be a zip_with operation?

Reply:
If this is acceptable, I'll run with it. Afterwards we can optimize the code further!